File: julia_backend.py
from PIL import Image
from matplotlib import cm
from itertools import product
import numpy as np
import julia_fast
import zmq
import io
import time

zmq_context = zmq.Context.instance()
recv_socket = zmq_context.socket(zmq.REP)
recv_socket.bind('tcp://127.0.0.1:5555')

def gen_image(key, w, h, cre, cim, cmap):
    start = time.perf_counter()
    m = julia_fast.julia_set(w, h, cre + cim*1j)
    end = time.perf_counter()-start
    
    image_data = np.empty((h,w,3), dtype=np.uint8)
    colors = 255*np.array(getattr(cm, cmap).colors)

    for j,i in product(range(h), range(w)):
        image_data[j,i,:] = colors[m[j,i]]

    image = Image.fromarray(image_data, mode='RGB')
    
    stream = io.BytesIO()
    image.save(stream, format='png')
    stream.seek(io.SEEK_SET)
    print('Made {} x {} image in {} seconds'.format(w, h, end))
    return stream.read()


if __name__ == '__main__':
    print('Listening to 127.0.0.1:5555')
    while True:
        message = recv_socket.recv_pyobj()
        recv_socket.send(gen_image(**message))

File: julia_backend.jl
using ZMQ
import JSON

ctx = Context()
recv_socket = Socket(ctx, REQ)
ZMQ.connect(recv_socket, "tcp://127.0.0.1:5556")

function julia_iteration(z, c, maxiter::Int64)
    for n = 1:maxiter
        if abs(z) > 2
            return n-1
        end
        z = z^2 + c
    end
    return maxiter
end

function julia_set(w::Int64, h::Int64, c, maxiter::Int64)
    i::Int64
    j::Int64
    m = Array(Int64, h, w)
    for j=1:h, i=1:w
        z = complex((i-w/2)/(h/2), (j-h/2)/(h/2))
	m[j,i] = julia_iteration(z, c, 256)
    end
    return m
end

function gen_image(key::AbstractString, w::Int64, h::Int64, cre::Float64, cri::Float64, cmap::AbstractString)
    tStart = time()

    tStop = time()
    println("Made ", w, " x ", h, " image in ", tStop, " seconds")
    return
